platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/builder/PlatformComponentBuilder.java

```diff
@@ -0,0 +1,214 @@
+ *         (with a very small number of exceptions due to tech debt that has not yet been paid off).</li>
+ *     <li>A component should have an interface and at default implementation.</li>
+ *     <li>A component should use {@link com.swirlds.common.wiring.component.ComponentWiring ComponentWiring} to define
+ *         wiring API.</li>
+ *     <li>The order in which components are constructed should not matter.</li>
+ * </ul>
+ */
+public class PlatformComponentBuilder {
+
+    private final PlatformBuildingBlocks blocks;
+
+    private EventHasher eventHasher;
+    private InternalEventValidator internalEventValidator;
+    private EventDeduplicator eventDeduplicator;
+
+    /**
+     * False if this builder has not yet been used to build a platform (or platform component builder), true if it has.
+     */
+    private boolean used;
+
+    /**
+     * Constructor.
+     *
+     * @param blocks the build context for the platform under construction, contains all data needed to
+     *                               construct platform components
+     */
+    public PlatformComponentBuilder(@NonNull final PlatformBuildingBlocks blocks) {
+        this.blocks = Objects.requireNonNull(blocks);
+    }
+
+    /**
+     * Get the build context for this platform. Contains all data needed to construct platform components.
+     *
+     * @return the build context
+     */
+    @NonNull
+    public PlatformBuildingBlocks getBuildingBlocks() {
+        return blocks;
+    }
+
+    /**
+     * Throw an exception if this builder has been used to build a platform or a platform factory.
+     */
+    private void throwIfAlreadyUsed() {
+        if (used) {
+            throw new IllegalStateException("PlatformBuilder has already been used");
+        }
+    }
+
+    /**
+     * Build the platform.
+     *
+     * @return the platform
+     */
+    @NonNull
+    public Platform build() {
+        throwIfAlreadyUsed();
+        used = true;
+
+        try (final ReservedSignedState initialState = blocks.initialState()) {
+            return new SwirldsPlatform(this);
+        } finally {
+
+            // Future work: eliminate the static variables that require this code to exist
+            if (blocks.firstPlatform()) {
+                MetricsDocUtils.writeMetricsDocumentToFile(
+                        getGlobalMetrics(),
+                        getPlatforms(),
+                        blocks.platformContext().getConfiguration());
+                getMetricsProvider().start();
+            }
+        }
+    }
+
+    /**
+     * Provide an event hasher in place of the platform's default event hasher.
+     *
+     * @param eventHasher the event hasher to use
+     * @return this builder
+     */
+    @NonNull
+    public PlatformComponentBuilder withEventHasher(@NonNull final EventHasher eventHasher) {
+        throwIfAlreadyUsed();
+        this.eventHasher = Objects.requireNonNull(eventHasher);
```

Comment:
Consider throwing if `this.eventHasher` is already defined - or can you think of a valid reason to call this multiple times?

(comment also applies to `withInternalEventValidator`, and `withEventDeduplicator`)

Reply:
Done.